### PR TITLE
Daniel to main

### PR DESCRIPTION
**Pull request by YoungMasterGandalf**

### CODE REVIEW

File: pgwidget/pgwidget_core.py [hunk 1]
```diff
@@ -1490,16 +1544,14 @@ def blit_text(self, surface, text, color=(0,0,0)):
...
 
-        for label in self.labels:
-            if label.is_cursor_drawing:
-                label._draw_cursor(screen)
-        #self.label.draw = self.blit_text
-        #self.label.draw(screen, self.label.text)
+        if auto_draw_cursor:
+            for label in self.labels:
+                if label.is_cursor_drawing:
+                    label._draw_cursor(screen)
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py:1549]: auto_draw_cursor - please add description what this means

Comment (DovaX) [reply pgwidget/pgwidget_core.py:1549]: docstring

File: pgwidget/pgwidget_core.py [hunk 2]
```diff
@@ -414,12 +455,18 @@ def on_key_down(self,event):
         
     def _move_one_row_up(self):
         self.shown_cursor_row_index = max(self.shown_cursor_row_index - 1, 0)
-        self.cursor_row_index = self.shown_cursor_row_index
+        # self.cursor_row_index = self.shown_cursor_row_index
+        self.cursor_row_index = self.shown_cursor_row_index + self.shown_text_first_line_index
+        if self.shown_cursor_offset_index > len(self.shown_text_rows[self.shown_cursor_row_index]):
+            self.shown_cursor_offset_index = len(self.shown_text_rows[self.shown_cursor_row_index])
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: change to camera offset

File: pgwidget/pgwidget_core.py [hunk 3]
```diff
@@ -414,12 +455,18 @@ def on_key_down(self,event):
...
 
     def _move_one_row_down(self):
         shown_text_rows = self.shown_text.split('\n')
         self.shown_cursor_row_index = min(self.shown_cursor_row_index + 1, len(shown_text_rows) - 1)
-        self.cursor_row_index = self.shown_cursor_row_index
+        # self.cursor_row_index = self.shown_cursor_row_index
+        self.cursor_row_index = self.shown_cursor_row_index + self.shown_text_first_line_index
+        if self.shown_cursor_offset_index > len(self.shown_text_rows[self.shown_cursor_row_index]):
+            self.shown_cursor_offset_index = len(self.shown_text_rows[self.shown_cursor_row_index])
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: change to camera offset

File: pgwidget/pgwidget_core.py [hunk 4]
```diff
@@ -361,7 +404,7 @@ def on_click(self,click_around_label_permitted=False,click_with_shift=False):
         if self.is_point_in_rectangle(pos) or click_around_label_permitted:
             self.shown_cursor_offset_index, self.shown_cursor_row_index = self._round_cursor_position_to_nearest_letter(pos)
             self.cursor_offset_index=self.shown_cursor_offset_index+self.shown_text_index_offset
-            self.cursor_row_index = self.shown_cursor_row_index
+            self.cursor_row_index = self.shown_cursor_row_index + self.shown_text_first_line_index
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: change to camera offset

File: pgwidget/pgwidget_core.py [hunk 5]
```diff
@@ -342,6 +383,8 @@ def _recalculate_cursor_position(self):
         if self.cursor_offset_index is not None or self.cursor_row_index is not None: #i.e. if cursor exists (it is not correct to put here shown_cursor_offset_index)
             text_length = self.get_text_pixel_length(self.shown_cursor_offset_index, self.shown_cursor_row_index)
             text_height = self.get_text_pixel_height(self.shown_cursor_row_index)
+            
+            # init_height = self.get_text_pixel_height(self.shown_text_first_line_index - 1) if self.shown_text_first_line_index > 0 else 0
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: remove

File: pgwidget/pgwidget_core.py [hunk 6]
```diff
@@ -282,8 +313,15 @@ def get_text_pixel_length(self,letter_index=None, letter_row = 0):
 
     def get_text_pixel_height(self, letter_row=0):
 
+        if self.shown_text:
+            # text_rows_heights = [self.myfont.size(text_line)[1] for text_line in self.shown_text[:letter_row + 1]]
+            text_rows_heights = [self.myfont.size('a')[1] for text_line in self.shown_text[:letter_row + 1]]
+            text_height = sum(text_rows_heights)
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: This will make performance dramatically slower - can we put there a constant or evaluate myfont.size just once?

Comment (DovaX) [reply pgwidget/pgwidget_core.py]: better to not use myfont.size at all

File: pgwidget/pgwidget_core.py [hunk 7]
```diff
@@ -191,6 +221,7 @@ def refresh_shown_text(self):
                 if len(row) > self.shown_text_max_length:
                     self.shown_text_rows[i] = row[:self.shown_text_max_length]
                     self.shown_text_rows[i+1] = row[self.shown_text_max_length:] + self.shown_text_rows[i+1]
+            self.shown_text_rows = self.shown_text_rows[self.shown_text_first_line_index: self.shown_text_last_line_index + 1]
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: replace with camera offset

File: pgwidget/pgwidget_core.py [hunk 8]
```diff
@@ -152,11 +181,12 @@ def cursor_row_index(self):
     def cursor_row_index(self,cursor_row_index):
         if cursor_row_index is not None:
             self._cursor_row_index = max(cursor_row_index, 0)
-            self.shown_cursor_row_index=self._cursor_row_index
+            self.shown_cursor_row_index = self._cursor_row_index - self.shown_text_first_line_index
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: replace with camera offset

File: pgwidget/pgwidget_core.py [hunk 9]
```diff
@@ -152,11 +181,12 @@ def cursor_row_index(self):
...
         else:
             self._cursor_row_index = None
         
-        self._recalculate_cursor_position() #TODO: add on_drag 
+        if self.shown_cursor_row_index is not None and 0 <= self.shown_cursor_row_index < len(self.shown_text_rows):
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: 0 <= self.shown_cursor_row_index < len(self.shown_text_rows)
????? what's wrong here? :)

Comment (DovaX) [reply pgwidget/pgwidget_core.py]: please also add a comment explaining the condition here

Comment (DovaX) [reply pgwidget/pgwidget_core.py]: do the same with other occurence of the same "if" above

File: pgwidget/pgwidget_core.py [hunk 10]
```diff
@@ -125,10 +147,15 @@ def text(self,text):
         if type(text)==str: #handling of NoneType
             if not self.is_multiline_label:
                 text = text.replace("\n", " ")
+
+        if self.is_multiline_label:
+            if len(self._text.split("\n")) != len(text.split("\n")):
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py:140]: potential problems on linux/mac ("\r\n" problem?)

File: pgwidget/pgwidget_core.py [hunk 11]
```diff
@@ -104,6 +107,25 @@ def __init__(self,text,color=(0,0,0),pos=[0,0],relative_pos=[0,0],font_type="Cal
...
+        self.refresh_shown_text()
+
+    @property
+    def shown_text_last_line_index(self):
+        return self._shown_text_last_line_index
+
+    @shown_text_last_line_index.setter
+    def shown_text_last_line_index(self, shown_text_last_line_index):
+        self._shown_text_last_line_index = shown_text_last_line_index
+        self.refresh_shown_text()
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: Discuss these two properties with dominik -> are they really needed?

File: pgwidget/pgwidget_core.py [hunk 12]
```diff
@@ -94,6 +94,9 @@ def __init__(self,text,color=(0,0,0),pos=[0,0],relative_pos=[0,0],font_type="Cal
         self.shown_text_index_offset=0 #how many letters is the shown text offset against original self.text
         self.shown_text=self.text
         self.shown_text_max_length=shown_text_max_length # in letters
+        self._shown_text_first_line_index = 0
+        self._shown_text_last_line_index = len(self.shown_text.split('\n'))
```
Comment (DovaX) [inline pgwidget/pgwidget_core.py]: are both needed?

Comment (DovaX) [reply pgwidget/pgwidget_core.py]: + replace with camera offset

**Review by DovaX — CHANGES_REQUESTED** [review]

File: pgwidget/pgwidget_core.py (hunk 10 above)
Comment (YoungMasterGandalf) [reply pgwidget/pgwidget_core.py:140]: Haven't encountered any on Linux.

**Review by DovaX — APPROVED** [review]

**Pull request merged**